### PR TITLE
Allow static checks to be run against an arbitrary path

### LINKED ISSUE #478: Allow static checks to be run against an arbitrary path
Currently the WP CLI command is limited to only run checks against plugins that are installed on the same site as Plugin Check. 

This is because it eventually calls the `Plugin_Request_Utility::get_plugin_basename_from_input()` method during the setup process which throws an Exception if the plugin is not installed.

To satisfy the requirements of https://github.com/WordPress/plugin-check/issues/441#issuecomment-2190961036, we need to be able to run checks against a plugin that is located outside the WordPress filesystem, e.g. `/tmp`.

Ideally, this would work by passing a CLI command like:

```
wp plugin check /tmp/plugin-path
```

When the path is not to an active plugin, only static checks should be eligible to be run.

### PR DESCRIPTION
**Pull request by davidperezgar**
What we need for the [submission form from wordpress.org](https://wordpress.org/plugins/developers/add/) is run static checks in an arbitrary path. It's helpful as well for any developer user to directly run static checks in any directory.

Resolves #478

### CODE REVIEW

**Comment by swissspidy** [comment]
I'm out of office until Monday but happy to review then.

At first glance I'd say the `is_dir` logic is at the wrong level. If you pull it higher up, then you don't to worry about naming or anything. For example doing that in `Abstract_Check_Runner::get_plugin_basename` would be better. No need to touch `get_plugin_basename_from_input()` at all.

Also, this would be an excellent candidate to write Behat tests for.

Happy to help with both once I'm back.

**Comment by ernilambar** [comment]
I tested PR and it is not working for plugin in external location as expected.
Few checks are working though.

In Plugin_Context there is `protected $main_file;` which is used in several checks. This expects the main file of the plugin but in external plugin, this variable just has plugin path. So there are few errors in the output.

Screenshot 2024-07-04 at 4 51 43 PM

Like in `Plugin_Header_Text_Domain_Check`, there is like this:
```
$plugin_main_file = WP_PLUGIN_DIR . '/' . $result->plugin()->basename();
```

May be we should fix `$main_file` in the beginning and access same property in all checks rather than evaluating main file like above.

**Comment by davidperezgar** [comment]
Thanks @ernilambar for your feedback. I've detected that problem, and I've committed a fix. Update the PR and it should work.

File: includes/Plugin_Context.php [hunk 1]
```diff
@@ -69,6 +69,26 @@ public function basename() {
...
 
+	/**
+	 * Returns the plugin main file.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @return string Plugin main file.
+	 */
+	public function main_file() {
+		if ( false === strpos( $this->main_file, '.php' ) ) {
```
Comment (ernilambar) [inline includes/Plugin_Context.php]: Finding actual main file here in this method is too late. May be it should be done in the constructor of `Plugin_Context`.

File: includes/Checker/Checks/Plugin_Readme_Check.php [hunk 2]
```diff
@@ -225,7 +225,7 @@ private function check_default_text( Check_Result $result, string $readme_file,
 	private function check_license( Check_Result $result, string $readme_file, Parser $parser ) {
 		$license          = $parser->license;
 		$matches_license  = array();
-		$plugin_main_file = WP_PLUGIN_DIR . '/' . $result->plugin()->basename();
+		$plugin_main_file = $result->plugin()->main_file();
```
Comment (ernilambar) [inline includes/Checker/Checks/Plugin_Readme_Check.php:231]: Similar update should be done in following files also.
- Plugin_Header_Text_Domain_Check
- Plugin_Readme_Check
- Plugin_Updater_Check
- Trademarks_Check

Comment (davidperezgar) [reply includes/Checker/Checks/Plugin_Readme_Check.php:231]: Yes, you're right. I'll check it out

File: includes/Plugin_Context.php (hunk 1 above)
Comment (davidperezgar) [reply includes/Plugin_Context.php]: ok, let's do it.

File: includes/Utilities/Plugin_Request_Utility.php [hunk 3]
```diff
@@ -40,12 +40,16 @@ class Plugin_Request_Utility {
...
-	 * @param string $plugin_slug The plugin slug or basename.
+	 * @param string $plugin_slug The plugin slug, basename or arbitrary path.
 	 * @return string The plugin basename.
 	 *
 	 * @throws Exception Thrown if an invalid basename or plugin slug is provided.
 	 */
 	public static function get_plugin_basename_from_input( $plugin_slug ) {
+		if ( is_dir( $plugin_slug ) ) {
+			return $plugin_slug;
+		}
```
Comment (swissspidy) [inline includes/Utilities/Plugin_Request_Utility.php]: We should not need to change this method.

This change should happen in `Abstract_Check_Runner::get_plugin_basename()`

Comment (davidperezgar) [reply includes/Utilities/Plugin_Request_Utility.php]: I don't understand. This conditional jump the test if the plugin is activated and get actived plugins, that is necessary.

Comment (swissspidy) [reply includes/Utilities/Plugin_Request_Utility.php]: This conditional should be done in `Abstract_Check_Runner::get_plugin_basename()`, not here.

Comment (ernilambar) [reply includes/Utilities/Plugin_Request_Utility.php]: May be like this in `Abstract_Check_Runner::get_plugin_basename()`:
```
			if ( is_dir( $plugin ) ) {
				$this->plugin_basename = $plugin;
			} else {
				$this->plugin_basename = Plugin_Request_Utility::get_plugin_basename_from_input( $plugin );
			}

```

File: includes/Utilities/Plugin_Request_Utility.php [hunk 4]
```diff
@@ -40,7 +40,7 @@ class Plugin_Request_Utility {
 	 *
 	 * @since 1.0.0
 	 *
-	 * @param string $plugin_slug The plugin slug or basename.
+	 * @param string $plugin_slug The plugin slug, basename or arbitrary path.
```
Comment (swissspidy) [inline includes/Utilities/Plugin_Request_Utility.php]:
```suggestion
	 * @param string $plugin_slug The plugin slug or basename.
```

**Review by swissspidy — APPROVED** [review]

**Review by ernilambar — APPROVED** [review]

File: includes/Plugin_Context.php [hunk 5]
```diff
@@ -69,6 +80,17 @@ public function basename() {
 		return plugin_basename( $this->main_file );
 	}
 
+	/**
+	 * Returns the plugin main file.
+	 *
+	 * @since 1.0.0
```
Comment (mukeshpanchal27) [inline includes/Plugin_Context.php]:
```suggestion
	 * @since 1.0.2
```

**Comment by davidperezgar** [comment]
None, it's working

**Pull request merged**